### PR TITLE
Refactor component update

### PR DESCRIPTION
**Pull request by ximex**
did some refactoring

### CODE REVIEW

**Comment by ximex** [comment]
@Fasse i think `$this->updateFinished` (ComponentUpdate) isn't needed any more, right?

**Comment by Fasse** [comment]
No, we don't need it.

**Comment by ximex** [comment]
strange. I only renamed the function and some variables. Please try again with this new commit and look if this log message with the same version (2.2.11) appears twice:
`'UPDATE: Start executing update steps to version '.$versionMain.'.'.$versionMinor.'.'.$tmpPatchVersion`

**Comment by Fasse** [comment]
now I got the following error if I open the update page:

> Notice: Undefined index: usf_name_intern in /admidio/adm_program/system/classes/profilefields.php on line 453
>
> Notice: Undefined index: usf_name_intern in /admidio/adm_program/system/classes/profilefields.php on line 457
>
> Fatal error: Uncaught UnexpectedValueException: Settings name "system_language" does not exist! in /admidio/adm_program/system/classes/settingsmanager.php:231 Stack trace: #0 /admidio/adm_program/system/classes/settingsmanager.php(309): SettingsManager->get('system_language', false) #1 /admidio/adm_program/installation/update.php(101): SettingsManager->getString('system_language') #2 {main} thrown in /admidio/adm_program/system/classes/settingsmanager.php on line 231

**Comment by ximex** [comment]
Did this only happen if you update from v2.x oder also if you update from v3.2? Or is it a special problem with PHP 5.3?

**Comment by Fasse** [comment]
I use PHP 5.6

I only get this if I update from 2.2 but at my first test I didn't get this error.

**Comment by Fasse** [comment]
hmm, strange the error is also in the master. 

But within version 3.2 an update from 2.2 is possible. Maybe the new SettingsManager is not tolerant enough.

**Comment by Fasse** [comment]
Ok, first test was with version 2.2 , error still exists.

Second test was with version 2.1, error still exists.

Both errors aren't available in version 3.2

**Comment by ximex** [comment]
arrr. this was a silly bug. now all ok?

File: adm_program/installation/update.php [hunk 1]
```diff
@@ -93,7 +94,7 @@
 $gProfileFields = new ProfileFields($gDb, $currOrgId);
```
Comment (Fasse) [inline adm_program/installation/update.php]: Why is this necessary?

This is the reason for the Notice if you have version 2.1 installed and perform an update:
> Notice: Undefined index: usf_name_intern in /Users/Markus/Sites/github/admidio/adm_program/system/classes/profilefields.php on line 457

If I comment this line update starts.

**Comment by Fasse** [comment]
Please have a look at my comment in the code.

Also after my fixes I know the problem of the update. If you start update from version 2.x the update for each sub version will perform multiple times. Thats the reason for the error:
> Notice: Undefined index: usf_name_intern in /Users/Markus/Sites/github/admidio/adm_program/system/classes/profilefields.php on line 457

File: adm_program/installation/update_functions.php [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+                require($phpUpdateFile);
+                $flagNextVersion = true;
+            }
+
+            $gLogger->notice('UPDATE: Finish executing update steps to version '.$versionMain.'.'.$versionMinor.'.'.$tmpPatchVersion);
+        }
+
+        // keine Datei mit der patch-version gefunden, dann die Main- oder minor-version hochsetzen,
+        // solange bis die aktuelle Versionsnummer erreicht wurde
+        if (!$flagNextVersion && version_compare($versionMain . '.' . $versionMinor . '.' . $tmpPatchVersion, ADMIDIO_VERSION, '<'))
```
Comment (Fasse) [inline adm_program/installation/update_functions.php]: Why do you only update $versionMinor if $flagNextVersion == false?
I think this is the reason for the loop within version 2.

Comment (ximex) [reply adm_program/installation/update_functions.php]: this is not my code. look here: https://github.com/Admidio/admidio/pull/687/files/7385e37afaa25d593ae971c012a29c52d8289d20#diff-c27e50483e21678ade212f20f1716125L422

File: adm_program/installation/update.php (hunk 1 above)
Comment (Fasse) [reply adm_program/installation/update.php]: @ximex what's about this?

File: adm_program/installation/update_functions.php (hunk 2 above)
Comment (Fasse) [reply adm_program/installation/update_functions.php]: hmm strange, within the new functions this won't work.

**Comment by Fasse** [comment]
Now everthing works.

**Comment by Fasse** [comment]
Now everthing works, you can merge.

File: adm_program/installation/db_scripts/update_3_3.xml [hunk 3]
```diff
@@ -4,6 +4,7 @@
     <step id="20">UPDATE %PREFIX%_user_fields SET usf_name = 'SYS_STREET' WHERE usf_name_intern = 'STREET' AND usf_name = 'SYS_ADDRESS'</step>
     <step id="30">UPDATE %PREFIX%_user_fields SET usf_name = 'SYS_STREET' WHERE usf_name_intern = 'STREET' AND usf_name = 'Adresse'</step>
     <step id="40">UPDATE %PREFIX%_user_fields SET usf_name = 'SYS_STREET' WHERE usf_name_intern = 'STREET' AND usf_name = 'Address'</step>
+    <step id="45">UPDATE %PREFIX%_preferences SET prf_name = 'dates_default_list_configuration' WHERE prf_name = 'dates_default_list_configurati'</step>
```
Comment (ximex) [inline adm_program/installation/db_scripts/update_3_3.xml]: @Fasse typo?
`'dates_default_list_configurati'`

File: adm_program/installation/db_scripts/update_3_3.xml [hunk 4]
```diff
@@ -43,24 +43,40 @@
...
     <step id="390">alter table %PREFIX%_roles_rights add constraint %PREFIX%_FK_ROR_ROR_PARENT foreign key (ror_ror_id_parent) references %PREFIX%_roles_rights (ror_id) on delete set null on update restrict</step>
-    <step id="400" database="pgsql">ALTER INDEX idx_%PREFIX%_mem_rol_usr_id RENAME TO %PREFIX%_idx_mem_rol_usr_id</step>
-    <step id="410" database="pgsql" error="ignore">ALTER INDEX idx_%PREFIX%_msc_part_id RENAME TO %PREFIX%_idx_msc_part_id</step>
-    <step id="420" database="pgsql" error="ignore">ALTER INDEX ak_%PREFIX%_shortname RENAME TO %PREFIX%_idx_org_shortname</step>
-    <step id="430" database="pgsql">ALTER INDEX idx_%PREFIX%_prf_org_id_name RENAME TO %PREFIX%_idx_prf_org_id_name</step>
+    <step id="395" database="pgsql">CREATE INDEX IF NOT EXISTS idx_%PREFIX%_msc_part_id ON %PREFIX%_messages_content (msc_part_id)</step>
+    <step id="397" database="pgsql">CREATE INDEX IF NOT EXISTS ak_%PREFIX%_shortname ON %PREFIX%_organizations (org_shortname)</step>
+    <step id="400" database="pgsql" error="ignore">ALTER INDEX idx_%PREFIX%_mem_rol_usr_id RENAME TO %PREFIX%_idx_mem_rol_usr_id</step>
```
Comment (ximex) [inline adm_program/installation/db_scripts/update_3_3.xml]: Why all this `error="ignore"`? What is the problem?

**Comment by ximex** [comment]
@Fasse please look at my two comments

File: adm_program/installation/db_scripts/update_3_3.xml (hunk 3 above)
Comment (Fasse) [reply adm_program/installation/db_scripts/update_3_3.xml]: no, this is a fix for an update from old versions, where prf_name length is only 30 chars

File: adm_program/installation/db_scripts/update_3_3.xml (hunk 4 above)
Comment (Fasse) [reply adm_program/installation/db_scripts/update_3_3.xml]: also problem of old admidio installations. There are some indices that are not in newer installations and within newer installations we got an error, but the update should not stop. Therefore ignore.
